---
title: sort-map-elements
---

# sort-map-elements

üíº This rule is enabled in the following [configs](/configs/): `recommended-alphabetical`, `recommended-line-length`, `recommended-natural`.

üîß This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

## üìñ Rule Details

Enforce sorted elements within a JavaScript `Map` object.

Sorting Map elements provides a clear and predictable structure to the codebase, making it easier for developers to locate and understand the key-value pairs defined within a Map.

This rule detects instances where Map elements are not sorted in a specified order and raises a linting error. It encourages developers to rearrange the elements in the desired order, ensuring a consistent structure across Map objects.

## üí° Examples

### Alphabetical and Natural Sorting

```ts
// Incorrect
let bebop = Map([
  ['spike', 'Spike Spiegel'],
  ['ed', 'Edward'],
  ['ein', 'Ein'],
  ['faye', 'Faye Valentine'],
  ['jet', 'Jet Black'],
])

// Correct
let bebop = Map([
  ['ed', 'Edward'],
  ['ein', 'Ein'],
  ['faye', 'Faye Valentine'],
  ['jet', 'Jet Black'],
  ['spike', 'Spike Spiegel'],
])
```

### Sorting by Line Length

```ts
// Incorrect
let bebop = Map([
  ['spike', 'Spike Spiegel'],
  ['ed', 'Edward'],
  ['ein', 'Ein'],
  ['faye', 'Faye Valentine'],
  ['jet', 'Jet Black'],
])

// Correct
let bebop = Map([
  ['spike', 'Spike Spiegel'],
  ['faye', 'Faye Valentine'],
  ['jet', 'Jet Black'],
  ['ed', 'Edward'],
  ['ein', 'Ein'],
])
```

## üîß Options

This rule accepts an options object with the following properties:

```ts
interface Options {
  type?: 'alphabetical' | 'natural' | 'natural'
  order?: 'asc' | 'desc'
  'ignore-case'?: boolean
}
```

### type

<sub>(default: `'alphabetical'`)</sub>

- `alphabetical` - sort alphabetically.
- `natural` - sort in natural order.
- `line-length` - sort by code line length.

### order

<sub>(default: `'asc'`)</sub>

- `asc` - enforce properties to be in ascending order.
- `desc` - enforce properties to be in descending order.

### ignore-case

<sub>(default: `false`)</sub>

Only affects alphabetical and natural sorting. When `true` the rule ignores the case-sensitivity of the order.

## ‚öôÔ∏è Usage

### Legacy Config

```json
// .eslintrc
{
  "rules": {
    "perfectionist/sort-map-elements": [
      "error",
      {
        "type": "line-length",
        "order": "desc"
      }
    ]
  }
}
```

### Flat Config

```js
// eslint.config.js
import perfectionist from 'eslint-plugin-perfectionist'

export default [
  {
    plugins: {
      perfectionist,
    },
    rules: {
      'perfectionist/sort-map-elements': [
        'error',
        {
          type: 'line-length',
          order: 'desc',
        },
      ],
    },
  },
]
```

## üöÄ Version

This rule was introduced in v0.5.0.

## üìö Resources

- [Rule source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/rules/sort-map-elements.ts)
- [Test source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/test/sort-map-elements.test.ts)
